File: BE_StayHub/app/Models/Region.php
<?php

namespace App\Models;

use App\Models\Concerns\HasUniqueSlug;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Laravel\Scout\Searchable;

class Region extends Model
{
    use HasFactory, HasUniqueSlug, Searchable;

    public const SEARCH_INDEX = 'regions';

    public function searchableAs(): string
    {
        return self::SEARCH_INDEX;
    }


    public const ACTIVE = true;
    public const INACTIVE = false;

    public const ACTIVE_LABELS = [
        self::ACTIVE => 'Đang hoạt động',
        self::INACTIVE => 'Ngừng hoạt động',
    ];

    protected $fillable = ['parent_id', 'code', 'name', 'path', 'slug', 'description', 'is_active', 'created_by'];

    protected function casts(): array
    {
        return ['is_active' => 'boolean'];
    }

    public function parent(): BelongsTo
    {
        return $this->belongsTo(Region::class, 'parent_id');
    }

    public function children(): HasMany
    {
        return $this->hasMany(Region::class, 'parent_id');
    }

    public function buildings(): HasMany
    {
        return $this->hasMany(Building::class);
    }

    public function creator(): BelongsTo
    {
        return $this->belongsTo(Admin::class, 'created_by');
    }

    /**
     * Dữ liệu đồng bộ sang Meilisearch để tra cứu khu vực tốc độ cao.
     */
    public function toSearchableArray(): array
    {
        return [
            'id' => $this->id,
            'parent_id' => $this->parent_id,
            'code' => $this->code,
            'name' => $this->name,
            'path' => $this->path,
            'slug' => $this->slug,
            'description' => $this->description,
            'is_active' => (bool) $this->is_active,
            'created_by' => $this->created_by,
            'created_at' => optional($this->created_at)->timestamp,
            'updated_at' => optional($this->updated_at)->timestamp,
        ];
    }
}

File: FE_StayHub_Mobile/lib/models/region.dart
class Region {
  final int id;
  final int? parentId;
  final String code;
  final String name;
  final String? path;
  final String? slug;
  final bool isActive;
  final String? description;
  final int? createdBy;
  final String? createdAt;
  final String? updatedAt;

  Region({
    required this.id,
    this.parentId,
    required this.code,
    required this.name,
    this.path,
    this.slug,
    required this.isActive,
    this.description,
    this.createdBy,
    this.createdAt,
    this.updatedAt,
  });

  int get status => isActive ? 1 : 2;

  factory Region.fromJson(Map<String, dynamic> json) {
    bool active = true;
    if (json['is_active'] != null) {
      active = json['is_active'] as bool;
    } else if (json['status'] != null) {
      active = (json['status'] as int) == 1;
    }

    return Region(
      id: json['id'] as int,
      parentId: json['parent_id'] as int?,
      code: json['code'] as String? ?? '',
      name: json['name'] as String? ?? '',
      path: json['path'] as String?,
      slug: json['slug'] as String?,
      isActive: active,
      description: json['description'] as String?,
      createdBy: json['created_by'] as int?,
      createdAt: json['created_at'] as String?,
      updatedAt: json['updated_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'parent_id': parentId,
      'code': code,
      'name': name,
      'path': path,
      'slug': slug,
      'is_active': isActive,
      'status': status,
      'description': description,
      'created_by': createdBy,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}
